Improving how chimeras are detected in CollectAlignmentSummaryMetrics.

nh13 opened the pull request:
Read pair orientation and SA tag are now considered when detecting
chimeras.

nh13 commented:
@yfarjoun @ktibbett thanks for your considering my request.

ktibbett commented:
@nh13 , do you have a sense of how much this changes results?

nh13 commented:
It changes the `PCT_CHIMERAS` significantly for samples that have a lot of chimeras.  I was actually a bit shocked that the presence of the "SA" tag was not being checked since that indicates it is a chimeric alignment.  

Nonetheless, on well-behaved samples the difference is small.  I took the first ~300K records from a [1000 Genomes Bam](http://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/working/20121126_NA12878_bam_downSampledTo5x/.).  The only metric that changed was `PCT_CHIMERAS` from `0.003719` to `0.003885` for the `PAIR` category.

nh13 commented:
Also, the build is breaking for a reason unknown to me and unrelated to this change.

ktibbett commented:
@eitanbanks , would you please take a look at this PR too?

eitanbanks commented on src/java/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -75,11 +79,13 @@
-        this.adapterKmers = prepareAdapterSequences();
-        this.maxInsertSize = maxInsertSize;
+                                            final boolean doRefMetrics, final List<String> adapterSequence, final int maxInsertSize,
+                                            PairOrientation expectedOrientation, boolean isBisulfiteSequenced) {
```
Please add 'final' (also to isBisulfiteSequenced once you're already here)

eitanbanks commented on src/java/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -318,11 +323,20 @@ else if(doRefMetrics) {
                                 if (Math.abs(record.getInferredInsertSize()) > maxInsertSize ||
-                                        !record.getReferenceIndex().equals(record.getMateReferenceIndex())) {
+                                        !record.getReferenceIndex().equals(record.getMateReferenceIndex()) ||
+                                        SamPairUtil.getPairOrientation(record) != expectedOrientation ||
```
I must admit that I don't know how exactly we define "chimeric" reads, but I just want to confirm with everyone:
this will count read pairs that overlap structural variants as chimeric.  E.g.  an inversion will cause reads to have unexpected orientations and now we're saying that those are chimeric.  (Although I suppose we might have already been doing that depending on the value for max insert size, as that could be capturing large deletions)

I would like @tfenne and @yfarjoun to confirm please.

tfenne replied:
@eitanbanks - yes, you are correct about SVs.  The problem is that, similar to trying to calculate base error rate, without going through a highly accurate variant calling process, it's impossible to tell apart read pairs that appear to come from different parts of the genome due to a biological structural variation, and read pairs that are the product of erroneous lab processes joining two or more disparate molecules together.

The same issue also existed with the prior definition.  As I see it @nh13's change takes us from detecting only anomalous inserts where the break is in between the two reads, to all anomalous inserts.

eitanbanks replied:
Okay, not objecting.  Just confirming.  Thanks!

eitanbanks commented on src/java/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -318,11 +323,20 @@ else if(doRefMetrics) {
+                            // Consider chimeras that occur *within* the read using the SA tag
+                            if (record.getMappingQuality() >= MAPPING_QUALITY_THRESOLD) {
+                                ++this.chimerasDenominator;
+                                if (record.getAttribute("SA") != null) ++ this.chimeras;
```
Please remove extra space for consistency.  :)

eitanbanks commented on src/java/picard/analysis/CollectAlignmentSummaryMetrics.java:
```diff
@@ -60,6 +61,13 @@
+ * Chimeras are identified if any of the of following criteria are met:
+ * <ul>
+ * <li>the insert size is larger than MAX_INSERT_SIZE</li>
+ * <li>the paired end orientation is different that the expected orientation</li>
```
"than" not that.
Also, please add: <li>the reads map to different contigs
(I'm just copying from the if statement in AlignmentSummaryMetricsCollector)

eitanbanks commented:
Okay, review done.  Nice!

nh13 commented:
@eitanbanks made the changes.  I didn't leave it as a second commit since they were trivial . I also rebased onto master.  Feel free to merge (previously travis was failing prior to the unit tests, not due to this commit, so perhaps someone should take a look).  Thanks for the review and accepting the changes.

eitanbanks commented:
@ktibbett am I allowed to merge?  Will let you do the honors when appropriate.

ktibbett commented:
@yfarjoun , want a quick look?

yfarjoun commented:
I'd like to look as well...tomorrow.



On Thu, Jan 14, 2016 at 5:43 PM, ktibbett <notifications@github.com> wrote:

> @yfarjoun <https://github.com/yfarjoun> , want a quick look?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/392#issuecomment-171805384>
> .
>

yfarjoun commented:
@nh13 can you rebase onto master as there was a problem with the travis build.

yfarjoun commented on src/java/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -318,11 +323,20 @@ else if(doRefMetrics) {
-                                        !record.getReferenceIndex().equals(record.getMateReferenceIndex())) {
+                                        !record.getReferenceIndex().equals(record.getMateReferenceIndex()) ||
+                                        SamPairUtil.getPairOrientation(record) != expectedOrientation ||
+                                        record.getAttribute("SA") != null) {
```
please use the enum SAMTag.SA

yfarjoun commented on src/java/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -318,11 +323,20 @@ else if(doRefMetrics) {
+                            // Consider chimeras that occur *within* the read using the SA tag
+                            if (record.getMappingQuality() >= MAPPING_QUALITY_THRESOLD) {
+                                ++this.chimerasDenominator;
+                                if (record.getAttribute("SA") != null) ++this.chimeras;
```
here too.

yfarjoun commented on src/java/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -318,11 +323,20 @@ else if(doRefMetrics) {
                                 }
                             }
                         }
+                        else { // fragment reads or read pairs with one end that maps
```
please comment what this is elseing about.

yfarjoun replied:
nm

yfarjoun commented:
other than my two small requests regarding the literal strings, I have a slight concern that the PCT_CHIMERAS isn't tested at all. I'm not sure what to say about this. on the one hand, I don't want to force you to add a test, on the other hand, I would like for there to be a test so we can see clearly what is and is not changing....

ideas?

nh13 commented:
@yfarjoun I totally understand and I'll leave that as a future improvement if you don't mind.  I am unable to make an issue in this project so I will leave it up to you.

I have fixed the one comment and rebased, so it should be ready to merge.

yfarjoun commented:
@ktibbett, woudl you be able to find someone who would add a test or two
that includes looking at the PCT_CHIMERAS (and have it equal something
other than zero...)?

On Fri, Jan 15, 2016 at 2:06 PM, Nils Homer <notifications@github.com>
wrote:

> @yfarjoun <https://github.com/yfarjoun> I totally understand and I'll
> leave that as a future improvement if you don't mind. I am unable to make
> an issue in this project so I will leave it up to you.
>
> I have fixed the one comment and rebased, so it should be ready to merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/392#issuecomment-172055574>
> .
>

ktibbett commented:
@yfarjoun , I will make an issue internally to create a test.  Merging this now.
